compiler/src/dmd/expressionsem.d

```diff
@@ -12789,14 +12789,23 @@ Expression dotIdSemanticProp(DotIdExp exp, Scope* sc, int flag)
         {
             if (exp.e1.isThisExp())
             {
+                //@@@DEPRECATED2.113@@@
+                // Delete the outermost if when turning into an error
+                deprecation(exp.loc, "`this` is only defined in non-static member functions, not `%s`", sc.parent.toChars());
```

Comment:
the deprecation message should be improved. I do not understand the "not '...'" at the end of it. Proposal:
`"'this' is not defined in static member functions. Use 'typeof(this)' or '%s'", ad.toChars()`

Reply:
The problem is that people use this in other static scopes, such as a struct scope.

Reply:
I wanted to also output the scope in which `this` is used. Take a look at the test cases where I had to modify the error. I think this is helpful, but if you insist I can rephrase, however, specifically mentioning "static member function" is not accurate.

Reply:
the message really is not understandable, so rephrasing is needed. "not Foo" is too brief. Perhaps "not inside scope Foo" would already help.